dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java

```diff
@@ -0,0 +1,270 @@
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.kernel.productinfo.ProductInfo;
+
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider;
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIServerXMLConfig;
+
+@Component(configurationPolicy = ConfigurationPolicy.OPTIONAL, configurationPid = "io.openliberty.microprofile.openapi")
+public class OpenAPIAppConfigProviderImpl implements OpenAPIAppConfigProvider {
+
+    private static final TraceComponent tc = Tr.register(OpenAPIAppConfigProviderImpl.class);
+
+    private static final String INVALID_APP_WARNING = "OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W";
+    private static final String INVALID_MODULE_WARNING = "OPEN_API_SLASH_IN_MODULE_CWWKO1679W";
+
+    private static final String INCLUDE_APP_PROPERTY_NAME = "includeApplication";
+    private static final String EXCLUDE_APP_PROPERTY_NAME = "excludeApplication";
+    private static final String INCLUDE_MODULE_PROPERTY_NAME = "includeModule";
+    private static final String EXCLUDE_MODULE_PROPERTY_NAME = "excludeModule";
+
+    private final List<OpenAPIAppConfigListener> openAPIAppConfigListeners = new ArrayList<OpenAPIAppConfigListener>();
+
+    private volatile MpConfigServerConfigObject config = null;
+
+    private boolean issuedBetaMessage;
+
+    @Activate
+    protected void activate(BundleContext context, Map<String, Object> properties) {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(this, tc, "Initial processing of server.xml");
+            }
+
+            MpConfigServerConfigObject newConfig = new MpConfigServerConfigObject(properties);
+            config = newConfig;
+        }
+    }
+
+    @Modified
+    protected void modified(Map<String, Object> properties) {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(this, tc, "Processing update to server.xml");
+            }
+
+            MpConfigServerConfigObject newConfig = new MpConfigServerConfigObject(properties);
+
+            if (config.equals(newConfig)) {
+                Tr.event(this, tc, "After update there is no need to rebuild the openAPI document");
+            } else {
+                Tr.event(this, tc, "Clearing out the outdated openAPI document");
+
+                config = newConfig;
+                openAPIAppConfigListeners.sort((OpenAPIAppConfigListener o1, OpenAPIAppConfigListener o2) -> Integer.compare(o1.getConfigListenerPriority(),
+                                                                                                                             o2.getConfigListenerPriority()));
+
+                for (OpenAPIAppConfigListener listener : openAPIAppConfigListeners) {
+                    listener.processConfigUpdate();
+                }
```

Comment:
Need to check what would happen if an exception were to be thrown here.

Reply:
Perhaps the question is what should happen. Lets say that DS handles it. If the first listener throws something that will prevent the second from running and, I think, leave us with the pre-updated documentation. Is that what we want?

Reply:
Hmm, I think we're ok with [the DS side](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.component.html#service.component-modified.method):
> If the modified method throws an exception, SCR must log an error message containing the exception with the Log Service, if present and continue processing the modification.

So hopefully, if there's a condition we haven't considered, the user gets an error message out that they can report.